src/citrine/informatics/constraints/ingredient_ratio_constraint.py

```diff
@@ -0,0 +1,107 @@
+    basis_ingredients: Optional[dict[str, float]]
+        map from ingredient to multiplier in the denominator of the ratio
+    basis_labels: Optional[dict[str, float]]
+        map from label to multiplier in the denominator of the ratio
+
+    """
+
+    formulation_descriptor = properties.Object(FormulationDescriptor, 'formulation_descriptor')
+    min = properties.Float('min')
+    max = properties.Float('max')
+    basis_ingredients = properties.Mapping(
+        properties.String, properties.Float, 'basis_ingredients', default={})
+    basis_labels = properties.Mapping(
+        properties.String, properties.Float, 'basis_labels', default={})
+
+    # The backend provides ingredients and labels as dictionaries, but presently only allows one
+    # between them. To clarify customer interaction, we only allow a single one of each to be set.
+    # Since our serde library doesn't allow extracting from a dict with unknown keys, we do it by
+    # hiding the dictionaries and exposing properties.
+    _ingredients = properties.Mapping(
+        properties.String, properties.Float, 'ingredients', default={})
+    _labels = properties.Mapping(properties.String, properties.Float, 'labels', default={})
+
+    typ = properties.String('type', default='IngredientRatio')
+
+    def __init__(self, *,
+                 formulation_descriptor: FormulationDescriptor,
+                 min: float,
+                 max: float,
+                 ingredient: Optional[Tuple[str, float]] = None,
+                 label: Optional[Tuple[str, float]] = None,
+                 basis_ingredients: Mapping[str, float] = {},
+                 basis_labels: Mapping[str, float] = {}):
+        self.formulation_descriptor = formulation_descriptor
+        self.min = min
+        self.max = max
+        self.ingredient = ingredient
+        self.label = label
+        self.basis_ingredients = basis_ingredients
+        self.basis_labels = basis_labels
+
+    @property
+    def ingredient(self):
+        """Retrieve the ingredient and its multiplier from the numerator, if it's been set."""
+        return self._numerator_read(self._ingredients)
+
+    @ingredient.setter
+    def ingredient(self, value: Optional[Tuple[str, float]]):
+        """Set the ingredient and its multiplier in the numerator."""
+        self._ingredients = self._numerator_validate(value, "Ingredient")
+
+    @property
+    def label(self):
+        """Retrieve the label and its multiplier from the numerator, if it's been set."""
+        return self._numerator_read(self._labels)
+
+    @label.setter
+    def label(self, value: Optional[Tuple[str, float]]):
+        """Set the label and its multiplier in the numerator."""
+        self._labels = self._numerator_validate(value, "Label")
+
+    def _numerator_read(self, num_dict):
+        if num_dict:
+            return tuple(num_dict.items())[0]
+        else:
+            return None
+
+    def _numerator_validate(self, value, input_type):
```

Comment:
The backend is also doing this validation logic, correct? I guess it's good to do it early, but perhaps not strictly necessary to have it client-side this things will eventually fall out of sync.

Reply:
_If_ we want to do more client-side validation, we may also want to validate that the basis is (1) has > 0 entries and (2) the basis multipliers are positive.

Reply:
Only sort of. The main purpose is to ensure they only provide a single key-value pair, since that's what we want to support in the client right now. Since there's no datatype in Python to let me restrict the length of a sequence (that I know of), I enforce it here.

I added the multiplier piece while I was here, but it could be removed to let the backend respond if we really want to. I'd just rather be strict in the client and relax those constraints later rather than the other way around.

Reply:
Sounds good, I won't complain about it.